pkg/operator/ceph/object/user/controller.go

```diff
@@ -0,0 +1,229 @@
+	return []metav1.OwnerReference{c.ownerRef}
+}
+
+func getObjectStoreUserObject(obj interface{}) (objectstoreuser *cephv1beta1.ObjectStoreUser, err error) {
+	var ok bool
+	objectstoreuser, ok = obj.(*cephv1beta1.ObjectStoreUser)
+	if ok {
+		// the objectstoreuser object is of the latest type, simply return it
+		return objectstoreuser.DeepCopy(), nil
+	}
+	return nil, fmt.Errorf("not a known objectstoreuser object: %+v", obj)
+}
+
+// Create the user
+func (c *ObjectStoreUserController) createUser(context *clusterd.Context, u *cephv1beta1.ObjectStoreUser) error {
+	// validate the user settings
+	if err := ValidateUser(context, u); err != nil {
+		return fmt.Errorf("invalid user %s arguments. %+v", u.Name, err)
+	}
+	//Set DisplayName to match Name if DisplayName is not set
+	displayName := u.Spec.DisplayName
+	if len(displayName) == 0 {
+		displayName = u.Name
+	}
+
+	// create the user
+	logger.Infof("creating user %s in namespace %s", u.Name, u.Namespace)
+	userConfig := cephrgw.ObjectUser{
+		UserID:      u.Name,
+		DisplayName: &displayName,
+	}
+	objContext := cephrgw.NewContext(context, u.Spec.Store, u.Namespace)
+
+	user, rgwerr, err := cephrgw.CreateUser(objContext, userConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create user %s. RadosGW returned error %d: %+v", u.Name, rgwerr, err)
+	}
+
+	// Store the keys in a secret
+	secrets := map[string]string{
+		"AccessKey": *user.AccessKey,
+		"SecretKey": *user.SecretKey,
+	}
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "rook-ceph-object-user-" + u.Spec.Store + "-" + u.Name,
+			Namespace: u.Namespace,
+			Labels: map[string]string{
+				"app":               AppName,
+				"user":              u.Name,
+				"rook_cluster":      u.Namespace,
+				"rook_object_store": u.Spec.Store,
+			},
+		},
+		StringData: secrets,
+		Type:       k8sutil.RookType,
+	}
+	k8sutil.SetOwnerRef(context.Clientset, u.Namespace, &secret.ObjectMeta, &c.ownerRef)
+
+	_, err = context.Clientset.CoreV1().Secrets(u.Namespace).Create(secret)
+	if err != nil {
+		return fmt.Errorf("failed to save user %s secret. %+v", u.Name, err)
+	}
+	logger.Infof("created user %s", u.Name)
+	return nil
+}
+
+// Delete the user
+
+// func deleteUser(context *clusterd.Context, u *cephv1beta1.ObjectStoreUser) error {
+// 	objContext := cephrgw.NewContext(context, u.Spec.Store, u.Namespace)
+// 	result, rgwerr, err := cephrgw.DeleteUser(objContext, u.Name)
+// 	if err != nil {
+// 		return fmt.Errorf("failed to delete user '%s'. RadosGW returned error %d: %+v", u.Name, rgwerr, err)
+// 	} else if result != "" {
+// 		logger.Infof("result of user delete is: %s", result)
+// 	}
+
+func deleteUser(context *clusterd.Context, u *cephv1beta1.ObjectStoreUser) error {
+	objContext := cephrgw.NewContext(context, u.Spec.Store, u.Namespace)
+	result, rgwerr, err := cephrgw.DeleteUser(objContext, u.Name)
+	if err != nil {
+		return fmt.Errorf("failed to delete user '%s': %+v", u.Name, err)
+	}
+	if rgwerr == 3 {
```

Comment:
If the DeleteUser function of the ceph rgw client tries to delete a user that does not exist, it returns ("", 3, failed to delete user: (passes up underlying error)). Given the err return was more dynamic than the static integer in the second return field, it was seemingly easier to key off that second field. 

I can key off the third field (err) if that is desired. Let me know how you would like to proceed.

Reply:
if `err != nil`, the error will be returned and you will never hit this check for 3, right? so it looks like you need to move this check inside the `err != nil` block above.

Reply:
Ah correct. I see the poor error handling now. I'll get it corrected. Thanks.

Reply:
I believe this should have what we are looking for: 
`func deleteUser(context *clusterd.Context, u *cephv1beta1.ObjectStoreUser) error {
	objContext := cephrgw.NewContext(context, u.Spec.Store, u.Namespace)
	result, _, err := cephrgw.DeleteUser(objContext, u.Name)
	if err != nil {
		if rgwerr == 3 {
			logger.Infof("user %s does not exist in store %s", u.Name, u.Spec.Store)
		}
		return fmt.Errorf("failed to delete user '%s': %+v", u.Name, err)
	}`

If err is not nil, and rgwerr is 3 (does not exist) then it should just log the informational and move on to cleaning up the secret. Otherwise it will throw the err if rgwerr != 3.

Reply:
do you need an `else` to return the error if the rgwerr isn't 3?

Reply:
You're right. I had mistakenly thought the if statement found rwgerr == 3 to be true, it would stop after logging the informational, but there is nothing to stop it from also returning the error below. Thanks.